pkg/api/client/client.go

```diff
@@ -0,0 +1,183 @@
+
+func WithAPIKey(key string) Option {
+	return func(c *DefaultClient) {
+		if key != "" {
+			c.key = key
+		}
+	}
+}
+
+func WithScheme(scheme string) Option {
+	return func(c *DefaultClient) {
+		if scheme != "" {
+			c.scheme = scheme
+		}
+	}
+}
+
+func WithName(name string) Option {
+	return func(c *DefaultClient) {
+		if name != "" {
+			c.name = fmt.Sprintf("%s/%s", name, version.Short())
+		}
+	}
+}
+
+const (
+	defaultRequestTimeout = 30 * time.Second
+	defaultClientName     = "default-client"
+)
+
+// NewDefault creates a new default client that can be used for all calls to goProbe APIs
+func NewDefault(addr string, opts ...Option) *DefaultClient {
+	c := &DefaultClient{
+		client:   http.DefaultClient,
+		scheme:   "http",
+		hostAddr: addr,
+		timeout:  defaultRequestTimeout,
+		name:     defaultClientName,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	c.client = &http.Client{
+		Transport: &transport{
+			rt: otelhttp.NewTransport(
+				http.DefaultTransport,
+			),
+			requestLogging: c.requestLogging,
+			clientName:     c.name,
+		},
+	}
+	return c
+}
+
+func (c *DefaultClient) Client() *http.Client {
+	return c.client
+}
+
+type transport struct {
+	rt             http.RoundTripper
+	requestLogging bool
+	clientName     string
+}
+
+func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
+	start := time.Now()
+
+	r.Header.Set("User-Agent", t.clientName)
+
+	// make request
+	resp, err := t.rt.RoundTrip(r)
+	duration := time.Since(start)
+
+	if t.requestLogging {
+		logger := logging.WithContext(r.Context()).With("req", slog.GroupValue(
+			slog.String("method", r.Method),
+			slog.String("url", r.URL.String()),
+			slog.String("user-agent", r.UserAgent()),
+			slog.Duration("duration", duration),
+		))
+		// log trace ID if it is present
+		sc := trace.SpanContextFromContext(r.Context())
+		if sc.HasTraceID() {
+			logger = logger.With(slog.String("traceID", sc.TraceID().String()))
+		}
+
+		if err == nil {
+			if resp != nil {
+				logger = logger.With("resp", slog.GroupValue(
+					slog.Int("status_code", resp.StatusCode),
+				))
+				switch {
+				case 200 <= resp.StatusCode && resp.StatusCode < 300:
+					logger.Info("completed request")
+				case 300 <= resp.StatusCode && resp.StatusCode < 400:
+					logger.Info("further action needed to complete request")
+				default:
+					logger.Error("server error returned")
+				}
+			} else {
+				logger.Error("empty response")
+			}
+		} else {
+			logger.Errorf("failed to send request: %v", err)
+		}
+	}
+	return resp, err
+}
+
+func (c *DefaultClient) Modify(ctx context.Context, req *httpc.Request) *httpc.Request {
+	// TODO: have a long, hard look at this function
+	//
+	// It did happen to suck the bytes out of the body, leaving nothing for the
+	// parser following it
+	//
+	// DO NOT uncomment before clarifying behavior with httpc library
+	//
+	// retry any request that isn't 2xx
+	// req = req.RetryBackOffErrFn(func(resp *http.Response, _ error) bool {
```

Comment:
I've also raised https://github.com/fako1024/httpc/issues/20 to cover the case (and maybe others) where the caller might miss the call to `RetryBackOff(...)` needed for `RetryBackOffErrFn(...)` to do anything.

Reply:
I was not aware of that at all, even though you mentioned the intervals before... Cost me half a day 🙈 . I guess it does make sense to couple them in a way so that the user knows to specify.

I'll adapt this then. Thanks for the clarification.

On a related note: being able to specify the type of backoff may help. E.g. stuff like `RetryExponentialBackOff`.

Reply:
Yeah, sorry about that. That's what I get if only a select few people use the package (and those happen to be the ones setting the requirements)... Your feedback is super helpful though, I'd like to feed back as much as possible into the core functionality.
To that point: I like the idea about providing "back-off strategies". However, I'd like to avoid pumping up the (already large) API even more with (potentially more confusing) methods. But I think what could work is to provide generator functions that can be used to created intervals based on the strategy and which can subsequently be put _into_ the method argument, like so:
```
.RetryBackOff(httpc.IntervalsExponential(nIntervals, intervalLength))
.RetryBackOff(httpc.IntervalsLinear(nIntervals, intervalLength))
```
What do you think?

Reply:
Awesome. I'll create an issue for that and take care. Since it doesn't break the API it's gonna be a small extension anyway!

Reply:
And sidenote: PR for the error handling is out as well: https://github.com/fako1024/httpc/pull/23 - I'll wait for approval on that one from someone at Nect though. I _assume_ it's used correctly everywhere, but just in case I don't want them to hit a jack-in-the-box when they next update the dependencies...